File: packages/agent/agent/lib/proactivity.ts
import type { Database } from "@clara-financas/db";
import { concepts } from "@clara-financas/db/schema/knowledge";
import { and, eq } from "drizzle-orm";

/**
 * O consentimento de proatividade — onde ele mora e como se lê.
 *
 * O princípio do produto diz "proatividade depende de relevância E
 * consentimento", e só a metade relevância existia: a varredura diária corria
 * para todo tenant `ready`, sem nenhuma porta para desligar os avisos como um
 * todo. (Cada compromisso individual sempre passou por aprovação — o que
 * faltava era o interruptor geral.)
 *
 * O estado vive num CONCEITO do bundle `learnings` — dado do próprio tenant,
 * atrás da RLS, com trilha de revisões — e não numa coluna do registry de
 * tenants: o registry é o mapa do control plane, e uma preferência da pessoa
 * é conteúdo, não mapa. Ausência de conceito = ligado, porque tudo que já
 * dispara aviso hoje foi individualmente aprovado num cartão.
 */
export const PROACTIVITY_CONCEPT_ID = "preferences/proactivity";

type Tx = Parameters<Parameters<Database["transaction"]>[0]>[0];

/** Lê o consentimento dentro de um escopo `forTenant` já aberto. */
export async function proactivityEnabled(tx: Tx, tenantId: string): Promise<boolean> {
  const [row] = await tx
    .select({ frontmatter: concepts.frontmatter })
    .from(concepts)
    .where(
      and(
        eq(concepts.tenantId, tenantId),
        eq(concepts.bundle, "learnings"),
        eq(concepts.conceptId, PROACTIVITY_CONCEPT_ID),
      ),
    )
    .limit(1);

  return row?.frontmatter?.enabled !== false;
}

File: services/agent/clara/knowledge/proactivity.py
"""O consentimento de proatividade — porta de `agent/lib/proactivity.ts`.

O princípio do produto diz "proatividade depende de relevância E
consentimento". Cada compromisso individual sempre passa por aprovação — o
que faltava era o interruptor geral. O estado vive num CONCEITO do bundle
`learnings` — dado do próprio tenant, atrás da RLS, com trilha de revisões —
e não numa coluna do registro de tenants: uma preferência da pessoa é
conteúdo, não mapa. Ausência de conceito = ligado, porque tudo que já
dispara aviso hoje foi individualmente aprovado num cartão.
"""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.orm import Session

from clara.db.models import Concept

PROACTIVITY_CONCEPT_ID = "preferences/proactivity"


def proactivity_enabled(session: Session, tenant_id: str) -> bool:
    frontmatter = session.execute(
        select(Concept.frontmatter).where(
            Concept.tenant_id == tenant_id,
            Concept.bundle == "learnings",
            Concept.concept_id == PROACTIVITY_CONCEPT_ID,
        )
    ).scalar_one_or_none()

    return frontmatter is None or frontmatter.get("enabled") is not False
